apiserver/Makefile

```diff
@@ -19,27 +42,16 @@ jupyter_notebook_config.py:
 	    -e "s,@keyfile@,$(GSA_KEY_FILE),g" \
             < jupyter_notebook_config.py.in > $@
 
-build: install-hail-locally jupyter_notebook_config.py
-	docker build -t spark-base -f Dockerfile.spark-base .
-	docker build -t hail-base -f Dockerfile.hail-base .
-	docker build -t spark-master -f Dockerfile.master .
-	docker build -t spark-worker -f Dockerfile.worker .
-	docker build -t apiserver -f Dockerfile.apiserver .
-	docker build -t hail-jupyter -f Dockerfile.hail-jupyter .
-
-push: SPARK_MASTER_IMAGE="gcr.io/$(PROJECT)/spark-master:$(shell docker images -q --no-trunc spark-master | sed -e 's,[^:]*:,,')"
-push: SPARK_WORKER_IMAGE="gcr.io/$(PROJECT)/spark-worker:$(shell docker images -q --no-trunc spark-worker | sed -e 's,[^:]*:,,')"
-push: APISERVER_IMAGE="gcr.io/$(PROJECT)/apiserver:$(shell docker images -q --no-trunc apiserver | sed -e 's,[^:]*:,,')"
-push: HAIL_JUPYTER_IMAGE="gcr.io/$(PROJECT)/hail-jupyter:$(shell docker images -q --no-trunc hail-jupyter | sed -e 's,[^:]*:,,')"
-push: build
-	docker tag spark-master $(SPARK_MASTER_IMAGE)
-	docker push $(SPARK_MASTER_IMAGE)
-	docker tag spark-worker $(SPARK_WORKER_IMAGE)
-	docker push $(SPARK_WORKER_IMAGE)
-	docker tag apiserver $(APISERVER_IMAGE)
-	docker push $(APISERVER_IMAGE)
-	docker tag hail-jupyter $(HAIL_JUPYTER_IMAGE)
-	docker push $(HAIL_JUPYTER_IMAGE)
+$(eval $(call DOCKER_IMAGE,spark-base,install-hail-locally jupyter_notebook_config.py,SPARK_BASE_IMAGE))
```

Comment:
Is the 3rd parameter used?

Reply:
It appears you're just directly constructing the target name (https://github.com/hail-is/hail/pull/5807/files#diff-61ca3ba8640a4ceacf10e37fce885bdfR24). If so, we can drop the 3rd argument.

Reply:
Ah, I was meant to define a variable called `SPARK_BASE_IMAGE` that was the fully qualified image name. I've fixed the definition of `DOCKER_IMAGE` to do so. I could also replace all instances of the `$$(shell docker ...)` with `$$($(3))` which will just reference the lazy variable, but I wasn't sure if that was more or less confusing than repeating the whole `shell docker ...` thing multiple times. The multiple levels of `$` escaping gets a bit hairy.

Reply:
Yeah its confusing, and also kind of neat